pkg/internal/apis/certmanager/validation/certificate.go

```diff
@@ -165,6 +177,24 @@ func validateUsages(a *internalcmapi.CertificateSpec, fldPath *field.Path) field
 	return el
 }
 
+func validateSecretTemplateLabels(crt *internalcmapi.CertificateSpec, fldPath *field.Path) field.ErrorList {
+	return metavalidation.ValidateLabels(crt.SecretTemplate.Labels, fldPath.Child("secretTemplate", "labels"))
+}
+
+func validateSecretTemplateAnnotations(crt *internalcmapi.CertificateSpec, fldPath *field.Path) field.ErrorList {
+	el := field.ErrorList{}
+
+	secretTemplateAnnotationsPath := fldPath.Child("secretTemplate", "annotations")
+	for a := range crt.SecretTemplate.Annotations {
+		if strings.HasPrefix(a, "cert-manager.io/") {
+			el = append(el, field.Invalid(secretTemplateAnnotationsPath, a, "cert-manager.io/* annotations are not allowed"))
+		}
+	}
```

Comment:
I just noticed that we are iterating on a map here. Since the iterating order is random on Go maps, this means we end up with flakiness, e.g. in https://prow.build-infra.jetstack.net/view/gs/jetstack-logs/logs/ci-cert-manager-previous-previous-experimental/1439782520404054016 😅 

To reproduce:

```sh
$ bazel run //pkg/internal/apis/certmanager/validation:go_default_test
$ ./bazel-bin/pkg/internal/apis/certmanager/validation/go_default_test_/go_default_test -test.run "TestValidateCertificate/invalid_with_disallowed_'CertificateSecretTemplate'_annotations" -test.count=10
--- FAIL: TestValidateCertificate (0.00s)
    --- FAIL: TestValidateCertificate/invalid_with_disallowed_'CertificateSecretTemplate'_annotations (0.00s)
        certificate_test.go:747: Expected error spec.secretTemplate.annotations: Invalid value: "cert-manager.io/alt-names": cert-manager.io/* annotations are not allowed but got spec.secretTemplate.annotations: Invalid value: "cert-manager.io/certificate-name": cert-manager.io/* annotations are not allowed
        certificate_test.go:747: Expected error spec.secretTemplate.annotations: Invalid value: "cert-manager.io/certificate-name": cert-manager.io/* annotations are not allowed but got spec.secretTemplate.annotations: Invalid value: "cert-manager.io/alt-names": cert-manager.io/* annotations are not allowed
FAIL
```

We can fix the flakiness by expecting the errors [in the unit test](https://github.com/jetstack/cert-manager/blob/4cb1386552089d6a8744de27a850b9fb32d56332/pkg/internal/apis/certmanager/validation/certificate_test.go#L744-L749) to appear in any order.

Instead of

```go
for i, e := range errs {
	expectedErr := s.errs[i]
	if !reflect.DeepEqual(e, expectedErr) {
		t.Errorf("Expected error %v but got %v", expectedErr, e)
	}
}
```

we could use [`ElementsMatch`](https://pkg.go.dev/github.com/stretchr/testify/assert#Assertions.ElementsMatch) (from the testify module):

```go
assert.ElementsMatch(t, s.errs, errs)
```

Reply:
Oops, I just noticed that https://github.com/jetstack/cert-manager/pull/4365 fixes the flakiness issue!

The reason we have this flakiness is because we forgot to backport this fix to the branch `release-1.5`.